topology/dynamic.go

```diff
@@ -113,66 +143,28 @@ func (t *dynamicTopology) Close() {
 
 	t.closed = true
 
-	t.w.Close()
-	t.s.Close()
+	t.watch.Close()
+	t.watchable.Close()
 }
 
-func waitOnInit(t xwatch.Source, d time.Duration) error {
+func waitOnInit(w xwatch.Watch, d time.Duration) error {
 	if d <= 0 {
 		return nil
 	}
 	select {
-	case <-t.WaitInit():
+	case <-w.C():
 		return nil
 	case <-time.After(d):
 		return errInitTimeOut
 	}
 }
 
-// serviceTopologyInput implements xwatch.SourceInput with retry
-type serviceTopologyInput struct {
-	w       xwatch.Watch
-	retrier xretry.Retrier
-	hashGen sharding.HashGen
-	closed  bool
-	// NB(chaowang) the watch may be initiated and not get any updates on its channel any time soon,
-	// so we should try whatever value the watch has as first poll to avoid source init time out.
-	// If the value the watch has is invalid, the source is still not initiated
-	poll int
-}
-
-func newServiceTopologyInput(w xwatch.Watch, h sharding.HashGen, opts xretry.Options) xwatch.SourceInput {
-	return &serviceTopologyInput{w: w, hashGen: h, retrier: xretry.NewRetrier(opts)}
-}
-
-func (s *serviceTopologyInput) Poll() (interface{}, error) {
-	var m Map
-	if finalErr := s.retrier.Attempt(func() error {
-		if s.poll > 0 {
-			// for normal attempts (from 2nd attempt on), block on the notification channel
-			if _, ok := <-s.w.C(); !ok {
-				s.closed = true
-				return xwatch.ErrSourceClosed
-			}
-		}
-		s.poll++
-
-		var err error
-		m, err = newMapFromServiceInstances(s.w.Get(), s.hashGen)
-		return err
-	}); finalErr != nil {
-		return nil, finalErr
-	}
-
-	return m, nil
-}
-
-func newMapFromServiceInstances(data interface{}, hashGen sharding.HashGen) (Map, error) {
+func getMapFromUpdate(data interface{}, hashGen sharding.HashGen) (Map, error) {
 	service, ok := data.(services.Service)
```

Comment:
if data is nil, ok will be false here, unless you want a specific err for nil interface{}?